### PR TITLE
[WIP] feat: improve unit testing mechanism

### PR DESCRIPTION
**Pull request by DecFox**
This diff introduces a new mechanism for improving unit tests for the existing code. The main features of this PoC are:
- Introduce a parent interface to allow mocking concrete types
- Introduce `test$Func` fields in concrete types to mock functions
- Allow better error handling by defining the errors we expect. Further, also check the type of error we get (using `errors.Is`) while writing tests. 

This is part of https://github.com/l3af-project/l3afd/issues/223

### CODE REVIEW

File: kf/bpf.go [hunk 1]
```diff
@@ -38,6 +38,9 @@ import (
 var (
 	execCommand           = exec.Command
 	copyBufPool sync.Pool = sync.Pool{New: func() interface{} { return new(bytes.Buffer) }}
+
+	// errTypeAssertion
```
Comment (dthaler) [inline kf/bpf.go:42]:
```suggestion
```

File: kf/bpf.go [hunk 2]
```diff
@@ -206,6 +211,51 @@ func StopExternalRunningProcess(processName string) error {
...
+	return b.Program.Artifact
+}
+
+// MapName returns the map name for the program
+func (b *BPF) MapName() string {
+	return b.Program.MapName
+}
+
+// MapNamePth returns the map name path
+func (b *BPF) MapNamePth() string {
```
Comment (dthaler) [inline kf/bpf.go:230]: Why "Pth" instead of "Path"?

File: kf/bpf.go [hunk 3]
```diff
@@ -206,6 +211,51 @@ func StopExternalRunningProcess(processName string) error {
...
+func (b *BPF) PrevMapNamePth() string {
+	return b.PrevMapNamePath
+}
+
+// ProgId returns the program ID
+func (b *BPF) ProgId() int {
+	return b.ProgID
+}
+
+// SeqId returns the program ID
```
Comment (dthaler) [inline kf/bpf.go:244]: I don't think it returns the program ID, that's ProgId() above.

File: kf/bpfmap.go [hunk 4]
```diff
@@ -108,12 +111,18 @@ func (b *MetricsBPFMap) GetValue() float64 {
 		// We have observed in smaller configuration VM's, if we restart KF's
 		// Stale mapID's are reported, in such cases re-checking map id
 		log.Warn().Err(err).Msgf("GetValue : NewMapFromID failed ID %d, re-looking up of map id", b.MapID)
-		tmpBPF, err := b.BPFProg.GetBPFMap(b.Name)
+		tmp, err := b.BPFProg.GetBPFMap(b.Name)
 		if err != nil {
-			log.Warn().Err(err).Msgf("GetValue: Update new map ID %d", tmpBPF.MapID)
+			log.Warn().Err(err).Msgf("GetValue: Update new map ID")
```
Comment (dthaler) [inline kf/bpfmap.go:116]: Why remove the old MapID from the log?

File: kf/nfconfig.go [hunk 5]
```diff
@@ -42,9 +43,44 @@ type NFConfigs struct {
...
 
+	TestNewBpfProgram   func(ctx context.Context, program models.BPFProgram, conf *config.Config) models.BPF
+	TestLoadRootProgram func(ifaceName string, direction string, progType string, conf *config.Config) (models.BPF, error)
+
 	mu *sync.Mutex
 }
 
+var (
+	// errUnknownHostname indicates that the hostname is invalid
+	errUnknownHostName = errors.New("unknown hostname")
```
Comment (dthaler) [inline kf/nfconfig.go:54]: "unknown" and "invalid" mean different things.  If you mean unknown, then update line 53 to not use the word "invalid".  If you mean invalid, then update both lines to not use the word "unknown".

File: kf/nfconfig.go [hunk 6]
```diff
@@ -42,9 +43,44 @@ type NFConfigs struct {
...
+	// errStopFailure indicates that there was a failure while stopping a program
+	errStopFailure = errors.New("failed to stop program")
+
+	// errUpdateFdFailure indicates that there was a failure while updating the program fd
+	errUpdateFdFailure = errors.New("failed to update program fd")
+
+	// errAddProgramFailure indicates that there was a failure while adding programs
+	errAddProgramFailure = errors.New("failed to add program")
+
+	// errDisableLROFailure indicates that there was a failure while disabling LRO
```
Comment (dthaler) [inline kf/nfconfig.go:74]: What is LRO? Expand acronym.  Large receive offload?

File: kf/nfconfig.go [hunk 7]
```diff
@@ -42,9 +43,44 @@ type NFConfigs struct {
...
+	errDisableLROFailure = errors.New("failed to disable LRO")
+
+	// errDownloadFailure indicates that there was a failure while downloading the BPF
```
Comment (dthaler) [inline kf/nfconfig.go:77]: What does it mean to download "the BPF"?

File: kf/nfconfig.go [hunk 8]
```diff
@@ -1018,73 +1070,58 @@ func (c *NFConfigs) AddProgramWithoutChaining(ifaceName string, bpfProgs *models
...
+				return fmt.Errorf("failed to add %v due to existing program %v on iface %v direction %v", bpfProg.Name, prog.Name(), ifaceName, progType)
 			}
 		}
 	}
 	return nil
 }
 
 // AddProgramsOnInterface: AddProgramsOnInterface will add given ebpf programs on given interface
-func (c *NFConfigs) AddProgramsOnInterface(ifaceName, HostName string, bpfProgs *models.BPFPrograms) error {
+func (c *NFConfigs) addProgramsOnInterface(ifaceName, HostName string, bpfProgs *models.BPFPrograms) error {
```
Comment (dthaler) [inline kf/nfconfig.go:1109]: This changes the capitalization to no longer match the comment in 1108

File: kf/nfconfig.go [hunk 9]
```diff
@@ -1175,95 +1208,79 @@ func (c *NFConfigs) AddeBPFPrograms(bpfProgs []models.L3afBPFPrograms) error {
...
 // DeleteProgramsOnInterface : It will delete ebpf Programs on the given interface
-func (c *NFConfigs) DeleteProgramsOnInterface(ifaceName, HostName string, bpfProgs *models.BPFProgramNames) error {
+func (c *NFConfigs) deleteProgramsOnInterface(ifaceName, HostName string, bpfProgs *models.BPFProgramNames) error {
 	if HostName != c.HostName {
-		errOut := fmt.Errorf("provided bpf programs do not belong to this host")
-		log.Error().Err(errOut)
-		return errOut
+		return fmt.Errorf("%w: provided bpf programs do not belong to this host", errUnknownHostName)
+		// log.Error().Err(errOut)
+		// return errOut
```
Comment (dthaler) [inline kf/nfconfig.go:1215]: Clean up commented-out code before marking this PR ready for review

File: kf/bpf.go (hunk 2 above)
Comment (DecFox) [reply kf/bpf.go:230]: We can't have member functions and fields of the same name (go does not allow that). It does feel a little uncanny and we can probably use a better name.

File: kf/bpfmap.go (hunk 4 above)
Comment (DecFox) [reply kf/bpfmap.go:116]: This needs to be undone. Thanks for pointing this out

File: kf/nfconfig.go (hunk 8 above)
Comment (dthaler) [reply kf/nfconfig.go:1109]: My comment is mainly about the inconsistency between the comment and the function name.  I have no opinion on whether it should be capitalized or not, only that the two should be consistent, which they're currently not in this PR.

Comment (DecFox) [reply kf/nfconfig.go:1109]: Sorry, I missed:
>  match the comment in 1108
Thank you for pointing this out.

**Pull request closed without merging**